File: app/src/test/java/com/mss/imagesearcher/presenter/imagelist/ImageListPresenterTest.java
package com.mss.imagesearcher.presenter.imagelist;

import com.mss.imagesearcher.domain.ImageListModel;
import com.mss.imagesearcher.domain.models.ImageInfo;
import com.mss.imagesearcher.view.imagelist.IImageListViewHolder;
import com.mss.imagesearcher.view.imagelist.ImageListView;

import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Maybe;
import io.reactivex.android.plugins.RxAndroidPlugins;
import io.reactivex.schedulers.Schedulers;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class ImageListPresenterTest {

    @Mock
    private ImageListView imageListView;

    @Mock
    private ImageListModel model;

    private ImageListPresenter imageListPresenter;
    private List<ImageInfo> imageInfoList;

    @BeforeClass
    public static void setupClass() {
        RxAndroidPlugins.setInitMainThreadSchedulerHandler(__ -> Schedulers.trampoline());
    }

    @Before
    public void setUp() {
        imageListPresenter = spy(new ImageListPresenter(model));
        imageListPresenter.attachView(imageListView);
    }

    private void initImageInfoList() {
        imageInfoList = new ArrayList();
        for (int i = 0; i < 50; i++) {
            ImageInfo imageInfo = new ImageInfo();
            imageInfo.setId(i);
            imageInfo.setWebFormatURL("https:\\\\previewurl" + i);
            imageInfo.setLargeImageURL("https:\\\\largeimagewurl" + i);
            imageInfoList.add(imageInfo);
        }
        when(model.getImages()).thenReturn(imageInfoList);
    }

    @Test
    public void onItemClick_openDetailActivity() {
        imageListPresenter.onItemClick(1);
        verify(imageListView).openDetailActivity(1);
    }

    @Test
    public void getRvPresenter_returnNotNull() {
        assertNotNull(imageListPresenter.getRvPresenter());
    }

    @Test
    public void onSearchClick_showResultFromNetwork() {
        initImageInfoList();
        when(model.getImagesFromNetwork(anyString(), eq(1))).thenReturn(Maybe.just(imageInfoList));

        imageListPresenter.onSearchClick("one");

        verify(imageListView).showProgress(true);
        verify(model).getImagesFromNetwork("one", 1);
        verify(imageListView, times(2)).refreshImageList();
        verify(imageListView).showProgress(false);
    }

    @Test
    public void onSearchClick_queryReturnNetworkException_showNotFoundMessage() {
        when(model.getImages()).thenReturn(new ArrayList<>());
        when(model.getImagesFromNetwork(anyString(), eq(1))).thenReturn(Maybe.error(new Exception()));

        imageListPresenter.onSearchClick("one");

        verify(imageListView).showProgress(true);
        verify(model).getImagesFromNetwork("one", 1);
        verify(imageListView, times(1)).refreshImageList();
        verify(imageListView).showProgress(false);
        verify(imageListView).showNotFoundMessage();
    }

    @Test
    public void onNeedNextPage_showResultFromNetwork() {
        initImageInfoList();
        when(model.getImagesFromNetwork(anyString(), eq(1))).thenReturn(Maybe.just(imageInfoList));

        imageListPresenter.onNeedNextPage();

        verify(imageListView).showProgress(true);
        verify(model, times(0)).getImagesFromNetwork("one", 1);
        verify(imageListView).refreshImageList();
        verify(imageListView).showProgress(false);
    }

    @Test
    public void onRefresh_showResultFromNetwork() {
        initImageInfoList();
        when(model.getImagesFromNetwork(anyString(), eq(1))).thenReturn(Maybe.just(imageInfoList));

        imageListPresenter.onRefresh("one");

        verify(imageListView).showProgress(true);
        verify(model).getImagesFromNetwork("one", 1);
        verify(imageListView, times(2)).refreshImageList();
        verify(imageListView).showProgress(false);
    }

    @Test
    public void onCreate_InitAdMob() {
        imageListPresenter.onCreate();

        verify(imageListView).initAdMob();
    }

    @Test
    public void onPrivacyPolicyClick_ShowPrivacyPolicy() {
        imageListPresenter.onPrivacyPolicyClick();

        verify(imageListView).showPrivacyPolicy();
    }

    @Test
    public void RvPresenterBindView_showProgressSetImage() {
        initImageInfoList();
        IImageListViewHolder holder = mock(IImageListViewHolder.class);
        when(holder.getPos()).thenReturn(1);

        imageListPresenter.getRvPresenter().bindView(holder);

        verify(holder).showProgress(true);
        verify(holder).setImage(imageInfoList.get(1).getWebFormatURL());
    }

    @Test
    public void RvPresenterGetItemCount_returnImagesCount() {
        initImageInfoList();

        assertEquals(imageInfoList.size(), imageListPresenter.getRvPresenter().getItemCount());
    }

    @Test
    public void RvPresenterOnImageLoaded_stopProgress() {
        IImageListViewHolder holder = mock(IImageListViewHolder.class);

        imageListPresenter.getRvPresenter().onImageLoaded(holder);

        verify(holder).showProgress(false);
    }

    @Test
    public void RvPresenterOnImageLoadFailed_stopProgress() {
        IImageListViewHolder holder = mock(IImageListViewHolder.class);

        imageListPresenter.getRvPresenter().onImageLoadFailed(holder);

        verify(holder).showProgress(false);
    }
}

File: app/src/test/java/com/mss/imagesearcher/presenter/imagelist/ImageListPresenterTest.kt
package com.mss.imagesearcher.presenter.imagelist

import com.mss.imagesearcher.domain.ImageListModel
import com.mss.imagesearcher.domain.models.ImageInfo
import com.mss.imagesearcher.view.imagelist.IImageListViewHolder
import com.mss.imagesearcher.view.imagelist.ListFragmentView
import io.reactivex.Maybe
import io.reactivex.android.plugins.RxAndroidPlugins
import io.reactivex.schedulers.Schedulers
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Before
import org.junit.BeforeClass
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.anyString
import org.mockito.ArgumentMatchers.eq
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.junit.MockitoJUnitRunner
import java.util.*

@RunWith(MockitoJUnitRunner::class)
class ImageListPresenterTest {

    @Mock
    private val listFragmentView: ListFragmentView? = null

    @Mock
    private val model: ImageListModel? = null

    private var imageListPresenter: ImageListPresenter? = null
    private var imageInfoList: MutableList<ImageInfo>? = null

    @Before
    fun setUp() {
        imageListPresenter = spy(ImageListPresenter(model!!))
        imageListPresenter!!.attachView(listFragmentView)
    }

    private fun initImageInfoList() {
        imageInfoList = ArrayList()
        for (i in 0..49) {
            val imageInfo = ImageInfo()
            imageInfo.id = i.toLong()
            imageInfo.webFormatURL = "https:\\\\previewurl$i"
            imageInfo.largeImageURL = "https:\\\\largeimagewurl$i"
            imageInfoList!!.add(imageInfo)
        }
        `when`(model!!.images).thenReturn(imageInfoList)
    }

    @Test
    fun onItemClick_openDetailActivity() {
        imageListPresenter!!.onItemClick(1)
        verify<ListFragmentView>(listFragmentView).openDetailActivity(1)
    }

    @Test
    fun getRvPresenter_returnNotNull() {
        assertNotNull(imageListPresenter!!.rvPresenter)
    }

    @Test
    fun onSearchClick_showResultFromNetwork() {
        initImageInfoList()
        `when`(model!!.getImagesFromNetwork(anyString(), eq(1))).thenReturn(Maybe.just(imageInfoList!!))

        imageListPresenter!!.onSearchClick("one")

        verify<ListFragmentView>(listFragmentView).showProgress(true)
        verify(model).getImagesFromNetwork("one", 1)
        verify<ListFragmentView>(listFragmentView, times(2)).refreshImageList()
        verify<ListFragmentView>(listFragmentView).showProgress(false)
    }

    @Test
    fun onSearchClick_queryReturnNetworkException_showNotFoundMessage() {
        `when`(model!!.images).thenReturn(ArrayList())
        `when`(model.getImagesFromNetwork(anyString(), eq(1))).thenReturn(Maybe.error(Exception()))

        imageListPresenter!!.onSearchClick("one")

        verify<ListFragmentView>(listFragmentView).showProgress(true)
        verify(model).getImagesFromNetwork("one", 1)
        verify<ListFragmentView>(listFragmentView, times(1)).refreshImageList()
        verify<ListFragmentView>(listFragmentView).showProgress(false)
        verify<ListFragmentView>(listFragmentView).showNotFoundMessage()
    }

    @Test
    fun onNeedNextPage_showResultFromNetwork() {
        initImageInfoList()
        `when`(model!!.getImagesFromNetwork(anyString(), eq(1))).thenReturn(Maybe.just(imageInfoList!!))

        imageListPresenter!!.onNeedNextPage()

        verify<ListFragmentView>(listFragmentView).showProgress(true)
        verify(model, times(0)).getImagesFromNetwork("one", 1)
        verify<ListFragmentView>(listFragmentView).refreshImageList()
        verify<ListFragmentView>(listFragmentView).showProgress(false)
    }

    @Test
    fun onRefresh_showResultFromNetwork() {
        initImageInfoList()
        `when`(model!!.getImagesFromNetwork(anyString(), eq(1))).thenReturn(Maybe.just(imageInfoList!!))

        imageListPresenter!!.onRefresh("one")

        verify<ListFragmentView>(listFragmentView).showProgress(true)
        verify(model).getImagesFromNetwork("one", 1)
        verify<ListFragmentView>(listFragmentView, times(2)).refreshImageList()
        verify<ListFragmentView>(listFragmentView).showProgress(false)
    }

    @Test
    fun onCreate_InitAdMob() {
        imageListPresenter!!.onCreate()

        verify<ListFragmentView>(listFragmentView).initAdMob()
    }

    @Test
    fun onPrivacyPolicyClick_ShowPrivacyPolicy() {
        imageListPresenter!!.onPrivacyPolicyClick()

        verify<ListFragmentView>(listFragmentView).showPrivacyPolicy()
    }

    @Test
    fun RvPresenterBindView_showProgressSetImage() {
        initImageInfoList()
        val holder = mock(IImageListViewHolder::class.java)
        `when`(holder.pos).thenReturn(1)

        imageListPresenter!!.rvPresenter.bindView(holder)

        verify(holder).showProgress(true)
        verify(holder).setImage(imageInfoList!![1].webFormatURL!!)
    }

    @Test
    fun RvPresenterGetItemCount_returnImagesCount() {
        initImageInfoList()

        assertEquals(imageInfoList!!.size.toLong(), imageListPresenter!!.rvPresenter.itemCount.toLong())
    }

    @Test
    fun RvPresenterOnImageLoaded_stopProgress() {
        val holder = mock(IImageListViewHolder::class.java)

        imageListPresenter!!.rvPresenter.onImageLoaded(holder)

        verify(holder).showProgress(false)
    }

    @Test
    fun RvPresenterOnImageLoadFailed_stopProgress() {
        val holder = mock(IImageListViewHolder::class.java)

        imageListPresenter!!.rvPresenter.onImageLoadFailed(holder)

        verify(holder).showProgress(false)
    }

    companion object {

        @BeforeClass
        fun setupClass() {
            RxAndroidPlugins.setInitMainThreadSchedulerHandler { Schedulers.trampoline() }
        }
    }
}